Add socketpair support to proc_open

kooldev opened the pull request:
Adds support for socket pairs to `proc_open()`. This is a big improvement over pipes on Windows systems because sockets do support non-blocking mode and can participate in `stream_select()`. Async PHP frameworks like ReactPHP and Amp have a hard time supporting process spawning on Windows because the created process pipes cannot be polled for events and therefore cannot be integrated with an event loop. Another common problem on Windows is that reads from process pipes will block and it is not always possible to alternate between reading data from a child processes STDOUT / STDERR as it arrives. In some cases this leads to a deadlock situation where a read blocks on one of the pipes and the child process fills the other pipes buffer without a way to drain it. Some libraries (Symfony Process Component for example) try to work around this by redirecting output to files and doing redirection in a shell.

Most of these problems can easily be solved by using a pair of connected sockets. The process calling `prop_open()` retains one of the sockets, the other one is passed to the child process as a `HANDLE` on Windows. This works only if the underlying socket does not make use of Overlapped mode, which is the default if a socket is created using `socket()`. We can make use of `WSASocket()` instead and not pass the `WSA_FLAG_OVERLAPPED` flag to create a socket that does not make use of overlapped IO. The PR uses a (slightly) modified copy of the `socketpair()` function implementation found in `win32/sockets.c` to achieve this. The change to the function could also be applied to the original implementation instead. This could be a BC break (overlapped mode dropped) but PHP does not use overlapped socket IO anywhere internally.

```php
$spec = [
    0 => ['null'],
    1 => ['socket'], // ['pipe', 'w']
    2 => ['socket'] // ['pipe', 'w']
];

$proc = proc_open(['ping', 'github.com'], $spec, $pipes);

foreach ($pipes as $pipe) {
    var_dump(stream_set_blocking($pipe, false));
}

$buffers = array_fill_keys(array_keys($pipes), '');

while ($pipes) {
    $r = $pipes;
    $w = null;
    $e = null;

    if (!stream_select($r, $w, $e, null, 0)) {
        throw new \Error("Select failed");
    }
    
    foreach ($r as $i => $pipe) {
        if (!is_resource($pipe) || feof($pipe)) {
            $line = trim($buffers[$i]);
            
            if ($line !== '') {
                echo "PIPE {$i} << {$line}\n";
            }
            
            echo "DROP pipe {$i}\n";
            unset($pipes[$i]);
            continue;
        }
        
        $chunk = @fread($pipe, 8192);
        
        if ($chunk === false) {
            throw new \Error("Failed to read: " . (error_get_last()['message'] ?? 'N/A'));
        }
        
        $buffers[$i] .= $chunk;
        
        if (false !== strpos($buffers[$i], "\n")) {
            $lines = array_map('trim', explode("\n", $buffers[$i]));
            $buffers[$i] = array_pop($lines);
            
            foreach ($lines as $line) {
                echo "PIPE {$i} << {$line}\n";
            }
        }
    }
}
```
The example demonstrates the use of socketpairs with a spawned process. It will show "realtime" data on Windows if sockets are used. If I use pipes instead it will block after the first read and then echo the remaining output once the `ping` process has finished. This approach might not work for every process but since use of sockets is an opt-in behavior ist should not break any existing code.

nikic commented:
> Most of these problems can easily be solved by using a pair of connected sockets. The process calling `prop_open()` retains one of the sockets, the other one is passed to the child process as a `HANDLE` on Windows. This works only if the underlying socket does not make use of Overlapped mode, which is the default if a socket is created using `socket()`. We can make use of `WSASocket()` instead and not pass the `WSA_FLAG_OVERLAPPED` flag to create a socket that does not make use of overlapped IO. The PR uses a (slightly) modified copy of the `socketpair()` function implementation found in `win32/sockets.c` to achieve this. The change to the function could also be applied to the original implementation instead. This could be a BC break (overlapped mode dropped) but PHP does not use overlapped socket IO anywhere internally.

@cmb69 Do you see any problem with modifying our `socketpair()` polyfill in this way? It would certainly be nicer to not duplicate the implementation.

kooldev commented:
@nikic One thing to watch out for is that only one of the sockets is opened in non-overlapped mode (sock[1]). The reasoning behind this is that it it changes the socket that is passed to the child while the one used in the parent process is identical to what socketpair() usually returns. It is possible to create a socketpair with both ends not using overlapped IO, but it requires using `AcceptEx()` instead of `accept()`. Usage is very different from standard accept and it has to be called through a function pointer. In addition to that it requires IOCP. I [did that first](https://github.com/kooldev/php-src/commit/526337b1af0055a2985f4ab9c36817e828727a46) but went with the copy for now because it is easier to use.

kooldev commented on ext/standard/proc_open.c:
```diff
@@ -451,7 +451,7 @@ static inline HANDLE dup_fd_as_handle(int fd)
 typedef struct _descriptorspec_item {
 	int index;                       /* desired FD # in child process */
-	int is_pipe;
+	int is_pipe;                     /* =1 for pipe / tty, =2 for socketpair */
```
Iintroduced an enum and renamed `is_pipe` to `type`.

```c
typedef enum _descriptor_type {
	DESCRIPTOR_TYPE_STD,
	DESCRIPTOR_TYPE_PIPE,
	DESCRIPTOR_TYPE_SOCKET
} descriptor_type;
```

cmb69 commented:
Very nice! That should indeed solve quite some issues regarding the always blocking pipes on Windows.

We should also think about AF_UNIX which is [available on recent Windows versions](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/), but this PR is good without that.

kooldev commented:
I did some more testing and all programs that do blocking IO worked just fine with sockets. I tried starting a node.js process with sockets and it does not work. I was expecting this because libuv does some magic to detect the pipe type and (re)open it in non-blocking mode. They do not cover the SOCKET case because it is uncommon and instead try to reopen as (named) pipe to read and write with IOCP. You will get output like this if you try:

```
PIPE 2 << net.js:327
PIPE 2 << err = this._handle.open(fd);
PIPE 2 << ^
PIPE 2 <<
PIPE 2 << Error: EINVAL: invalid argument, uv_pipe_open
PIPE 2 << at new Socket (net.js:327:26)
PIPE 2 << at createWritableStdioStream (internal/bootstrap/switches/is_main_thread.js:67:18)
PIPE 2 << at process.getStdout [as stdout] (internal/bootstrap/switches/is_main_thread.js:117:12)
PIPE 2 << at Object.get (internal/console/constructor.js:171:38)
PIPE 2 << at Object.Console.<computed> (internal/console/constructor.js:298:46)
PIPE 2 << at Object.log (internal/console/constructor.js:309:61)
PIPE 2 << at Object.<anonymous> (C:\php-sdk\examples\console.js:15:9)
PIPE 2 << at Module._compile (internal/modules/cjs/loader.js:1138:30)
PIPE 2 << at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
PIPE 2 << at Module.load (internal/modules/cjs/loader.js:986:32) {
PIPE 2 << errno: -4071,
PIPE 2 << code: 'EINVAL',
PIPE 2 << syscall: 'uv_pipe_open'
PIPE 2 << }
```
I think that this is not much of a problem, you can always use a real pipe or redirect to a file instead by passing this to the spec argument of `proc_open()`. Maybe it can even be supported in node.js / libuv but that is not related to PHP at all. I also did some testing with .bat files that prompt for user input and a socket as STDIN, it works just fine.

nikic commented on ext/standard/proc_open.c:
```diff
@@ -725,6 +732,41 @@ static int set_proc_descriptor_to_pipe(descriptorspec_item *desc, zend_string *z
+
+	if (create_socketpair(sock)) {
+		zend_string *err = php_socket_error_str(php_socket_errno());
+		php_error_docref(NULL, E_WARNING, "Unable to create socket pair %s", ZSTR_VAL(err));
```
```suggestion
		php_error_docref(NULL, E_WARNING, "Unable to create socket pair: %s", ZSTR_VAL(err));
```

nikic commented on ext/standard/proc_open.c:
```diff
@@ -1203,38 +1248,46 @@ PHP_FUNCTION(proc_open)
-		if (descriptors[i].is_pipe) {
-			switch (descriptors[i].mode_flags) {
+		if (descriptors[i].type != DESCRIPTOR_TYPE_STD) {
+			if (descriptors[i].type == DESCRIPTOR_TYPE_PIPE) {
```
You should be able to move this code into the else branch below and reduce the scope of mode_string.

nikic commented on ext/standard/proc_open.c:
```diff
@@ -1203,38 +1248,46 @@ PHP_FUNCTION(proc_open)
-			fcntl(descriptors[i].parentend, F_SETFD, FD_CLOEXEC);
+				/* Mark the descriptor close-on-exec, so it won't be inherited by
+				 * potential other children */
+				fcntl(descriptors[i].parentend, F_SETFD, FD_CLOEXEC);
```
We have this code in two different places now for the socket and the pipe case. Can we keep it just here (outside the if/else)? (I'd also move the corresponding windows dup_handle code next to it, assuming I'm not missing some reason why these happen in different places.)

nikic commented on ext/standard/tests/general_functions/proc_open_sockets3.phpt:
```diff
@@ -0,0 +1,55 @@
+
+$proc = proc_open($cmd, $spec, $pipes);
+
+var_dump(stream_set_blocking(pipes[1], false));
```
```suggestion
var_dump(stream_set_blocking($pipes[1], false));
```

kooldev commented:
@nikic This should be ready to merge now. AppVeyor reports 3 failed tests, but they seem to be unrelated to the PR.

I added 2 more test cases that deal with STDIN being set to a socket or a pipe. I needed to make a little adjustment to `detect_is_seekable()` in `plain_wrapper.c` in order to make PHP properly read from a socket passed as STDIN. The Windows API function `GetFileType()` detects pipes, named pipes and sockets as `FILE_TYPE_PIPE` ([documented here](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfiletype)). I added a call to `GetNamedPipeInfo()` to check if the handle is actually a pipe (works with both anonymous and named pipes). If this fails `is_pipe` is reset to avoid calls to `PeekNamedPipe()` during read operations. Without this change the peek operation will fail (because it is not supported by sockets) and every read will result in an empty string being returned.

cmb69 commented:
Looks good to me. Thanks!

dktapps commented:
It seems this wasn't documented or mentioned in the 8.0 changelogs. I'm only just now learning about it.

cmb69 commented:
@dktapps, it's [mentioned in the migration guide](https://www.php.net/manual/en/migration80.new-features.php#migration80.new-features.standard); the actual `proc_open()` manpage still needs to be updated, though.
